src/worker/multithreaded.rs

```diff
@@ -40,25 +54,42 @@ impl Worker for MultiThreadedWorker {
         self.senders[component] = Some(sender);
         Ok(())
     }
+
     fn append_row(&mut self, row: (usize, Vec<i16>)) -> Result<()> {
         let component = row.0;
         let sender = mem::replace(&mut self.senders[component], None).unwrap();
         sender.send(WorkerMsg::AppendRow(row.1)).expect("jpeg-decoder worker thread error");
         self.senders[component] = Some(sender);
         Ok(())
     }
-    fn get_result(&mut self, index: usize) -> Result<Vec<u8>> {
+
+    fn get_result_with(
+        &mut self,
+        index: usize,
+        collect: impl FnOnce(Receiver<Vec<u8>>) -> Vec<u8>,
+    ) -> Result<Vec<u8>> {
         let (tx, rx) = mpsc::channel();
         let sender = mem::replace(&mut self.senders[index], None).unwrap();
         sender.send(WorkerMsg::GetResult(tx)).expect("jpeg-decoder worker thread error");
-        Ok(rx.recv().expect("jpeg-decoder worker thread error"))
+        Ok(collect(rx))
     }
 }
 
-fn spawn_worker_thread(component: usize) -> Result<Sender<WorkerMsg>> {
-    let (tx, rx) = mpsc::channel();
+impl Worker for StdThreadWorker {
+    fn start(&mut self, row_data: RowData) -> Result<()> {
+        self.0.start_with(row_data, spawn_worker_thread)
+    }
+    fn append_row(&mut self, row: (usize, Vec<i16>)) -> Result<()> {
+        self.0.append_row(row)
+    }
+    fn get_result(&mut self, index: usize) -> Result<Vec<u8>> {
+        self.0.get_result_with(index, collect_worker_thread)
+    }
+}
 
-    spawn(component, move || {
+fn create_worker() -> (Sender<WorkerMsg>, impl FnOnce() + 'static) {
+    let (tx, rx) = mpsc::channel();
```

Comment:
The fact that this is being done at all means rayon is being used as a pool of generic spawned threads. I can't see any advantages to using rayon over any other simple threadpool implementation here, which could do it with less overhead.

Reply:
We're not recreating threads, that's still a small advantage. But indeed, you're welcome to create a PR that implements the result passing in a proper manner otherwise. I, franlky, don't see how except by restructuring due to the absence of `JoinHandle` or other futures for tasks?

Reply:
Any generic threadpool implementation can avoid recreating threads the same way, rayon is a large dependency to pull in when it's only being used to spawn threads and join them at the end. Your existing std multi-threading code is probably marginally more efficient than this code.